cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/DistributedPreviewRunner.java

```diff
@@ -0,0 +1,124 @@
+import io.cdap.cdap.messaging.MessagingService;
+import io.cdap.cdap.proto.BasicThrowable;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.cdap.spi.data.StructuredTableAdmin;
+import io.cdap.cdap.spi.data.table.StructuredTableRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Distributed implementation of {@link PreviewRunner}.
+ */
+public class DistributedPreviewRunner extends DefaultPreviewRunner {
+  private static final Logger LOG = LoggerFactory.getLogger(DistributedPreviewRunner.class);
+  private static final Gson GSON = new Gson();
+
+  private final Path programIdFilePath;
+
+  @Inject
+  public DistributedPreviewRunner(
+    MessagingService messagingService,
+    DatasetOpExecutorService dsOpExecService,
+    DatasetService datasetService,
+    LogAppenderInitializer logAppenderInitializer,
+    ApplicationLifecycleService applicationLifecycleService,
+    ProgramRuntimeService programRuntimeService,
+    ProgramLifecycleService programLifecycleService,
+    PreviewDataPublisher previewDataPublisher,
+    DataTracerFactory dataTracerFactory,
+    NamespaceAdmin namespaceAdmin,
+    MetricsCollectionService metricsCollectionService,
+    ProgramNotificationSubscriberService programNotificationSubscriberService,
+    LevelDBTableService levelDBTableService,
+    StructuredTableAdmin structuredTableAdmin,
+    StructuredTableRegistry structuredTableRegistry,
+    CConfiguration cConf) {
+    super(messagingService, dsOpExecService, datasetService, logAppenderInitializer, applicationLifecycleService,
+          programRuntimeService, programLifecycleService, previewDataPublisher, dataTracerFactory, namespaceAdmin,
+          metricsCollectionService, programNotificationSubscriberService, levelDBTableService, structuredTableAdmin,
+          structuredTableRegistry);
+    this.programIdFilePath = Paths.get(cConf.get(Constants.CFG_LOCAL_DATA_DIR), "program.id").toAbsolutePath();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    super.startUp();
+
+    if (Files.exists(programIdFilePath)) {
+      try (Reader reader = Files.newBufferedReader(programIdFilePath, StandardCharsets.UTF_8)) {
+        ProgramId programId = GSON.fromJson(reader, ProgramId.class);
+        long submitTimeMillis = RunIds.getTime(programId.getApplication(), TimeUnit.MILLISECONDS);
+        PreviewStatus status = new PreviewStatus(
+          PreviewStatus.Status.RUN_FAILED, submitTimeMillis,
+          new BasicThrowable(new Exception("Preview runner container killed possibly because of out of memory. " +
+                                             "Please try running preview again.")),
+          null, null);
+        setStatus(programId, status);
+      }
+      destroy();
```

Comment:
Will calling destroy() transit this guava `Service` to COMPLETED state?? Do you want this service stop?

Reply:
No `destroy()` just deletes the file on line 116 here - https://github.com/cdapio/cdap/pull/12589/files/a9a6ec661469ccafa8e3aa3ee77b462040882b2b#diff-fe8b7d5449ae6f60f24802aa2104bd47R116

It wont cause any service state transition. 
Should I rename `init` to `onPreviewStart` and `destroy` to `onPreviewEnd` for clarity?